src/fs/metadata_log_bootstrap.hh

```diff
@@ -0,0 +1,485 @@
+    }
+
+    // Returns whether the reading was successful
+    bool read_sstring(sstring& str, size_t size) {
+        str.resize(size);
+        return read(str.data(), size);
+    }
+
+    std::optional<temporary_buffer<uint8_t>> read_tmp_buff(size_t size) {
+        if (_pos + size > _size) {
+            return std::nullopt;
+        }
+
+        _pos += size;
+        return temporary_buffer<uint8_t>(_data + _pos - size, size);
+    }
+
+    // Returns whether the processing was successful
+    bool process_crc_without_reading(boost::crc_32_type crc, size_t size) {
+        if (_pos + size > _size) {
+            return false;
+        }
+
+        crc.process_bytes(_data + _pos, size);
+        return true;
+    }
+
+    std::optional<data_reader> extract(size_t size) {
+        if (_pos + size > _size) {
+            return std::nullopt;
+        }
+
+        _pos += size;
+        return data_reader(_data + _pos - size, size);
+    }
+};
+
+class metadata_log_bootstrap {
+    metadata_log& _metadata_log;
+    cluster_range _available_clusters;
+    std::unordered_set<cluster_id_t> _taken_clusters;
+    std::optional<inode_t> _next_cluster;
+    temporary_buffer<uint8_t> _curr_cluster_data;
+    data_reader _curr_cluster;
+    data_reader _curr_checkpoint;
+
+    metadata_log_bootstrap(metadata_log& metadata_log, cluster_range available_clusters)
+    : _metadata_log(metadata_log)
+    , _available_clusters(available_clusters)
+    , _curr_cluster_data(decltype(_curr_cluster_data)::aligned(metadata_log._alignment, metadata_log._cluster_size))
+    {}
+
+    future<> bootstrap(cluster_id_t first_metadata_cluster_id, fs_shard_id_t fs_shards_pool_size, fs_shard_id_t fs_shard_id) {
+        _next_cluster = first_metadata_cluster_id;
+        return do_with((cluster_id_t)first_metadata_cluster_id, [this](cluster_id_t& last_cluster) {
+            return do_until([this] { return not _next_cluster.has_value(); }, [this, &last_cluster] {
+                cluster_id_t curr_cluster = *_next_cluster;
+                _next_cluster = std::nullopt;
+                bool inserted = _taken_clusters.emplace(curr_cluster).second;
+                assert(inserted); // TODO: check it in next_cluster record
+                last_cluster = curr_cluster;
+                return bootstrap_cluster(curr_cluster);
+            }).then([this, &last_cluster] {
+                // Initialize _curr_cluster_buff
+                _metadata_log._curr_cluster_buff = make_lw_shared<metadata_to_disk_buffer>(_metadata_log._cluster_size, _metadata_log._alignment, 0);
+                _metadata_log._curr_cluster_buff->reset_from_bootstraped_cluster(cluster_id_to_offset(last_cluster, _metadata_log._cluster_size), _curr_cluster_data.get(), _curr_cluster.curr_pos(), metadata_log::align_after_flush);
+            });
+        }).then([this, fs_shards_pool_size, fs_shard_id] {
+            // Initialize _cluser_allocator
+            std::deque<cluster_id_t> free_clusters;
```

Comment:
We need to add clusters with medium and big writes to _taken_clusters. I would add a TODO for that.